src/components/input.js

```diff
@@ -23,6 +23,50 @@ export const textInput = function(props) {
 }
 
 
+export const numberInput = props => {
+  return h(Interactive, _.mergeAll([{
+    as: 'input',
+    type: 'number',
+    style: {
+      width: '100%',
+      paddingLeft: '1rem',
+      paddingRight: '0.25rem',
+      fontWeight: 300,
+      fontSize: 14
+    }
+  }, Style.elements.input, props]))
+}
+
+
+export class IntegerInput extends Component {
+  constructor(props) {
+    super(props)
+    this.state = { textValue: undefined, lastValue: undefined }
+  }
+
+  static getDerivedStateFromProps({ value }, { lastValue }) {
```

Comment:
I haven't seen this before. I'm not completely following what this does.

Reply:
It's a newish react hook that's meant to replace the deprecated `componentWillReceiveProps`. The overall goal here is to have a 'semi-controlled' component. The final value is stored in the parent's state, but the input tracks changes locally as it's being edited. On blur, it sends the new value up to the parent. But we ultimately defer to the `value` prop, and if we ever see a change in that, we let it override the state.

The end result is that the component allows full editing, including through temporarily invalid states, but the parent only sees a series of valid values.